Panic when collecting boot logs for AzureMachine

/kind bug

[Before submitting an issue, have you checked the [Troubleshooting Guide](https://capz.sigs.k8s.io/topics/troubleshooting.html)?]

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
When "Collecting boot logs for AzureMachine", invalid memory address error happens.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x252b4c1]

goroutine 1 [running]:
sigs.k8s.io/cluster-api-provider-azure/test/e2e.collectVMBootLog({0x2d6e8c0, 0xc00012a008}, 0xc00039a000, {0xc00001c050, 0x4d})
	/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/azure_logcollector.go:363 +0xc1
```

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/test/e2e/azure_logcollector.go#L363

Log:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cloud-provider-azure/1841/pull-cloud-provider-azure-e2e-ccm-capz/1536926584823353344/build-log.txt

**What did you expect to happen:**
No error

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]


**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

--- a/test/e2e/azure_logcollector.go
+++ b/test/e2e/azure_logcollector.go
@@ -58,7 +58,7 @@ var _ framework.ClusterLogCollector = &AzureLogCollector{}
 
 // CollectMachineLog collects logs from a machine.
 func (k AzureLogCollector) CollectMachineLog(ctx context.Context, managementClusterClient client.Client, m *clusterv1.Machine, outputPath string) error {
-	var errors []error
+	var errs []error
 
 	am, err := getAzureMachine(ctx, managementClusterClient, m)
 	if err != nil {
@@ -73,19 +73,19 @@ func (k AzureLogCollector) CollectMachineLog(ctx context.Context, managementClus
 	hostname := getHostname(m, isAzureMachineWindows(am))
 
 	if err := collectLogsFromNode(ctx, managementClusterClient, cluster, hostname, isAzureMachineWindows(am), outputPath); err != nil {
-		errors = append(errors, err)
+		errs = append(errs, err)
 	}
 
 	if err := collectVMBootLog(ctx, am, outputPath); err != nil {
-		errors = append(errors, err)
+		errs = append(errs, errors.Wrap(err, "Unable to collect VM Boot Diagnostic logs"))
 	}
 
-	return kinderrors.NewAggregate(errors)
+	return kinderrors.NewAggregate(errs)
 }
 
 // CollectMachinePoolLog collects logs from a machine pool.
 func (k AzureLogCollector) CollectMachinePoolLog(ctx context.Context, managementClusterClient client.Client, mp *expv1.MachinePool, outputPath string) error {
-	var errors []error
+	var errs []error
 	var isWindows bool
 
 	am, err := getAzureMachinePool(ctx, managementClusterClient, mp)
@@ -112,11 +112,11 @@ func (k AzureLogCollector) CollectMachinePoolLog(ctx context.Context, management
 			hostname := mp.Status.NodeRefs[i].Name
 
 			if err := collectLogsFromNode(ctx, managementClusterClient, cluster, hostname, isWindows, filepath.Join(outputPath, hostname)); err != nil {
-				errors = append(errors, err)
+				errs = append(errs, err)
 			}
 
 			if err := collectVMSSBootLog(ctx, instance, filepath.Join(outputPath, hostname)); err != nil {
-				errors = append(errors, err)
+				errs = append(errs, errors.Wrap(err, "Unable to collect VMSS Boot Diagnostic logs"))
 			}
 		} else {
 			Logf("MachinePool instance %s does not have a corresponding NodeRef", instance)
@@ -124,7 +124,7 @@ func (k AzureLogCollector) CollectMachinePoolLog(ctx context.Context, management
 		}
 	}
 
-	return kinderrors.NewAggregate(errors)
+	return kinderrors.NewAggregate(errs)
 }
 
 // collectLogsFromNode collects logs from various sources by ssh'ing into the node
@@ -359,6 +359,10 @@ func windowsNetworkLogs(execToPathFn func(outputFileName string, command string,
 // collectVMBootLog collects boot logs of the vm by using azure boot diagnostics.
 func collectVMBootLog(ctx context.Context, am *v1beta1.AzureMachine, outputPath string) error {
 	Logf("Collecting boot logs for AzureMachine %s\n", am.GetName())
+
+	if am == nil || am.Spec.ProviderID == nil {
+		return errors.New("AzureMachine provider ID is nil")
+	}
 
 	resourceId := strings.TrimPrefix(*am.Spec.ProviderID, azure.ProviderIDPrefix)
 	resource, err := autorest.ParseResourceID(resourceId)

